### ISSUE TITLE
GVim on Linux doesn't dynamically adjust to user interface scaling

### ISSUE THREAD

**Issue by bokand** [vim/vim#2736]
I have a laptop with a High DPI monitor running Debian. When using the laptop screen, user interface scaling is set to "Double". When I plug into an external monitor, I switch this to "Normal". GVim looks correct when its started in either mode. However, switching from Normal to Double or vice versa while GVim is open causes the fonts to blow up in size (when going from Double->Normal) or shrink (in the other direction). Closing and reopening GVim reloads with the correct font size. Attached screenshots after switching from Double -> Normal and then after GVim is reopened.

It seems there's already support for correctly handing High DPI scenarios, it just needs to detect the situation change dynamically. Having to close GVim and reopen it everytime I switch modes is very inefficient since I lose all my context.

**Comment by bokand** [comment]
![screenshot from 2018-03-21 09-43-14](https://user-images.githubusercontent.com/11840257/37713277-c4dd0528-2cec-11e8-9edb-97a5a09a6d59.png)
![screenshot from 2018-03-21 09-42-14](https://user-images.githubusercontent.com/11840257/37713278-c4ed2656-2cec-11e8-9f8c-307ac0ca737d.png)

**Comment by tonymec** [comment]
I don't use external monitors; but no matter how I change the screen size (even by dragging the sides of an xterm with Vim inside) Vim notices it and adjusts its `'lines'` and `'columns'` settings. Of course, Vim running in an xterm has no control over the font, but it also notices the size change when I maximize or un-maximize the gvim screen by means of the buttons near top right of its titlebar.

My guess is that when you switch from Normal to Double or vice-versa, your screen interface does not properly notify gvim of the resizing event.

* Which gvim version are you using? (as of this writing, the latest one is 8.0.1625)
* Which GUI is your gvim compiled for? (GTK2, GTK3, …)
* Which system component defines your user interface scaling? (which commands and/or mouse clicks do you use to change it?)

Reminder: The more detailed a bug report the better (and see also "How to Ask Questions the Smart Way" http://catb.org/~esr/faqs/smart-questions.html).

Best regards,
Tony.

**Comment by bokand** [comment]
Thanks for the quick reply and apologies for lack of details.

The lines and cols are indeed correctly updated when I resize the window. However, the fonts don't scale with the updated UI scale factor. I'm changing the scale factor in the cinnamon-settings app (under General>User Interface Scaling). gvim should have full control over its font if I'm not mistaken. I think gvim is getting whatever resize/rescale signal is sent since the rest of the UI (i.e. the toolbars, titlebar, etc) elements are scaled correctly - though those look like they maybe common widgets that do that automatically. The content pane is the only thing that looks wrong.

Here's some more details, happy to provide more on request or do some diagnostics/log dives (I'm just not sure where to poke):

gvim version 8.0.1401 on Debian "buster" with Cinnamon. I believe it's compiled for GTK2, here's the compile flags if that helps:

Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0/ -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/
pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/freetype2 -Wdate-time  -g -O2 -fdebug-prefix-map=/tmp/build-debs.U8UMDD/vim-8.0.1401-2=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: gcc   -L. -Wl,-z,relro -Wl,-z,now -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E  -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -o vim   -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux  -lacl -lattr -lgpm -ldl  -L/usr/l
ib -llua5.2 -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/x86_64-linux-gnu/perl/5.24/CORE -lperl -ldl -lm -lpthread -lcrypt -L/usr/lib/python2.7/config-x86_64-linux-gnu -lpython2.7 -lpthread -ldl -lutil -lm -Xlinker -export-dynamic -Wl,-O1 -Wl,-Bsymbolic-functions  -L/usr/lib/x86_64-linux-gnu -ltcl8.6 -ldl -lz -lpthread -lieee -lm -lruby-2.3 -lpthread -lgmp -ldl -lcrypt -lm

**Comment by tonymec** [comment]
Yes, gvim has full control of its fonts, and yes, your gvim (like mine) is compiled with GTK2 GUI. In that gvim flavour, gvim sets exactly two font parameters, namely the font family name and the point size. For instance, my vimrc includes the following:
``` vim
if has('gui_running')
  set guifont=Bitstream\ Vera\ Sans\ Mono\ 8
endif
```
which means that I set the font name to "Bitstream Vera Sans Mono" and its size to 8pt.

About how gvim sets the font, and how you can influence it, see `:help 'guifont'` and http://vim.wikia.com/wiki/Setting_the_font_in_the_GUI

IIUC, the gvim code (i.e. the C code written by Bram et al. and distributed via a git server on github and via a Mercurial mirror on bitbucket) is not aware of the size of its titlebar and toolbar: that's part of what GTK and/or GDK do "behind the scenes". If the screen-resize event doesn't reach that gvim code it won't adjust how it draws the font, and in particular, if the gvim code doesn't know that the DPI has changed it won't change the number of pixels it uses vertically and horizontally for each "character box".

Maybe someone who knows Debian and Cinnamon better than I do will be able to help you better than I can.

Best regards,
Tony.

**Comment by bokand** [comment]
Thank you Tony. Yes, it seems the font needs to be updated for DPI in response to the resize. This already happens during initialization so I think it's just a matter of refreshing on some event. If someone who knows about Cinnamon and gvim code could point me at the right window manager event to listen to and a starting point in gvim for setting font size I could possibly have a go at building gvim mysef.

**Comment by tonymec** [comment]
About setting font size, see my previous post.

About what to listen to, you might try setting an autocommand on the `VimResized` event (q.v.) but I haven't tried it (and anyway, I'm on openSUSE, not Debian, and I don't use Cinnamon).

Here's a possible (untested) snippet for your vimrc. If it doesn't work, well, try something else :-/
``` vim
if has('gui_running') && has('autocmd')
  au VimEnter * let s:fontsetting = &guifont    " save the font at the end of startup
  au VimResized * let &guifont = s:fontsetting  " set the font again if the GUI changes size
endif
```


Best regards,
Tony.

**Comment by nuko8** [comment]
(For those who prefer a hard way:)

> If someone who knows about Cinnamon and gvim code could point me at the right window manager event to listen to and a starting point in gvim for setting font size I could possibly have a go at building gvim mysef.

`GdkScreen` supports two different signals, `monitor-changed` and `size-changed`. By connecting  a GdkScreen instance to user defined functions via `g_signal_connect()` function family, one could set the resolution for font handling on the screen using `gdk_screen_set_resolution()` there.  (I don't know about Cinnamon at all, though.)
